Add mysensors tcp ethernet gateway

MartinHjelmare opened the pull request:
**Description:**

**This will BREAK existing configs for mysensors.** Change "- port:" under "gateways" to "- device:". See below for new example config.
- Bump version of pymysensors to 0.6, which includes the tcp gateway.
- Update requirements_all.txt.
- Replace CONF_PORT with CONF_DEVICE and ATTR_PORT with ATTR_DEVICE.
- Add tcp_port in config.
- Try to guess if tcp or serial gateway is configured, by validating
  device name as an ip address. If successful setup tcp gateway, if it
  fails, setup serial gateway.
- Update device_state_attributes to show correct device, ethernet or
  serial.

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
mysensors:
  gateways:
    - device: '/dev/ttyUSB0'
      persistence_file: 'path/mysensors.json'
      baud_rate: 38400
    - device: '/dev/ttyACM0'
      persistence_file: 'path/mysensors2.json'
      baud_rate: 115200
    - device: '192.168.1.18'
      persistence_file: 'path/mysensors3.json'
      tcp_port: 5003
  debug: true
  persistence: true
  version: '1.5'
  optimistic: false
```

**Checklist:**

If code communicates with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [x] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [x] New dependencies are only imported inside functions that use them ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

MartinHjelmare committed:
Add mysensors tcp ethernet gateway

* Bump version of pymysensors to 0.6, which includes the tcp gateway.
* Update requirements_all.txt.
* Replace CONF_PORT with CONF_DEVICE and ATTR_PORT with ATTR_DEVICE.
* Add tcp_port in config.
* Try to guess if tcp or serial gateway is configured, by validating
	device name as an ip address. If successful setup tcp gateway, if it
	fails, setup serial gateway.
* Update device_state_attributes to show correct device, ethernet or
	serial.

MartinHjelmare commented:
Local tox run is failing due to unrelated bluepy dependency right now.

balloob commented on homeassistant/components/binary_sensor/mysensors.py:
```diff
@@ -101,8 +101,16 @@ def name(self):
     @property
     def device_state_attributes(self):
         """Return device specific state attributes."""
+        try:
```
Instead of these `try … except` statements, use `hasattr(self.gateway, 'server_address')` or `getattr(self.gateway, 'server_address', None)`

MartinHjelmare replied:
Yes, that should work a lot cleaner.

balloob commented on homeassistant/components/sensor/mysensors.py:
```diff
@@ -157,8 +157,16 @@ def unit_of_measurement(self):
     @property
     def device_state_attributes(self):
         """Return device specific state attributes."""
+        try:
```
I see this piece of code copy/paste 3 times now. Should you make a helper?

MartinHjelmare replied:
Yes, good point.

MartinHjelmare committed:
Comment out bluepy and clean up attributes

* Add bluepy to commented requirements in gen_requirements_all.py and
	comment out the corresponding requirement in requirements_all.txt.
* Use getattr to get host "server_address" of ethernet gateway if it
	exists, and	use the address for the "device" attribute in
	device_state_attributes. If the gateway is not an ethernet gateway,
	getattr will return None and gateway.port will be used instead for
	device attribute.

MartinHjelmare commented:
Damnit... bluepy_devices used by eq3btsmart depends on bluepy, so now that can't import correctly. Did I do the right thing by commenting bluepy and running gen_requirements_all? Too tired right now to think properly. Hope we have time to fix this.

balloob commented:
Merged as 9d391be
